src/platform/android/device.rs

```diff
@@ -0,0 +1,246 @@
+        let fd = match config.raw_fd {
+            Some(raw_fd) => raw_fd,
+            _ => return Err(Error::InvalidConfig),
+        };
+        let device = {
+            let tun = Fd::new(fd).map_err(|_| io::Error::last_os_error())?;
+
+            Device {
+                queue: Queue { tun: tun },
+            }
+        };
+        Ok(device)
+    }
+
+    /// Split the interface into a `Reader` and `Writer`.
+    pub fn split(self) -> (posix::Reader, posix::Writer) {
+        let fd = Arc::new(self.queue.tun);
+        (posix::Reader(fd.clone()), posix::Writer(fd.clone()))
+    }
+
+    /// Return whether the device has packet information
+    pub fn has_packet_information(&self) -> bool {
+        self.queue.has_packet_information()
+    }
+
+    #[cfg(feature = "mio")]
+    pub fn set_nonblock(&self) -> io::Result<()> {
+        self.queue.set_nonblock()
+    }
+}
+
+impl Read for Device {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.queue.tun.read(buf)
+    }
+}
+
+impl Write for Device {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.queue.tun.write(buf)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.queue.tun.flush()
+    }
+}
+
+impl D for Device {
+    type Queue = Queue;
+
+    fn name(&self) -> &str {
+        return "";
+    }
+
+    fn set_name(&mut self, value: &str) -> Result<()> {
+        Err(Error::NotImplemented)
+    }
+
+    fn enabled(&mut self, value: bool) -> Result<()> {
+        Ok(())
+    }
+
+    fn address(&self) -> Result<Ipv4Addr> {
+        Err(Error::NotImplemented)
+    }
+
+    fn set_address(&mut self, value: Ipv4Addr) -> Result<()> {
+        Ok(())
+    }
+
+    fn destination(&self) -> Result<Ipv4Addr> {
+        Err(Error::NotImplemented)
+    }
+
+    fn set_destination(&mut self, value: Ipv4Addr) -> Result<()> {
+        Ok(())
+    }
+
+    fn broadcast(&self) -> Result<Ipv4Addr> {
+        Err(Error::NotImplemented)
+    }
+
+    fn set_broadcast(&mut self, value: Ipv4Addr) -> Result<()> {
+        Ok(())
+    }
+
+    fn netmask(&self) -> Result<Ipv4Addr> {
+        Err(Error::NotImplemented)
+    }
+
+    fn set_netmask(&mut self, value: Ipv4Addr) -> Result<()> {
+        Ok(())
+    }
+
+    fn mtu(&self) -> Result<i32> {
+        Err(Error::NotImplemented)
+    }
+
+    fn set_mtu(&mut self, value: i32) -> Result<()> {
+        Ok(())
+    }
+
+    fn queue(&mut self, index: usize) -> Option<&mut Self::Queue> {
+        if index > 0 {
+            return None;
+        }
+
+        Some(&mut self.queue)
+    }
+}
+
+impl AsRawFd for Device {
+    fn as_raw_fd(&self) -> RawFd {
+        self.queue.tun.as_raw_fd()
+    }
+}
+
+impl IntoRawFd for Device {
+    fn into_raw_fd(self) -> RawFd {
+        self.queue.tun.into_raw_fd()
+    }
+}
+
+pub struct Queue {
+    tun: Fd,
+}
+
+impl Queue {
+    pub fn has_packet_information(&self) -> bool {
+        // on ios this is always the case
```

Comment:
Is this also the case for Android?

Reply:
I just simply copy-and-paste... Need some time to look into it.

Reply:
https://developer.android.com/reference/android/net/VpnService.Builder#establish()

Reading the fd will retrieve a raw IP packet without 4 bytes header. So I think this function should always return a `false`.

Reply:
Sounds correct, I'm fine with merging this after that's fixed and you can make another pull request if you find stuff that doesn't work properly on Android later on :)